src/manifest/manifest.ts

```diff
@@ -46,426 +52,130 @@ const generateSupplementaryTrackID = idGenerator();
+    removeEventListener: eventEmitter.removeEventListener.bind(eventEmitter),
   };
 
-  /**
-   * Construct a Manifest instance from a parsed Manifest object (as returned by
-   * Manifest parsers) and options.
-   *
-   * Some minor errors can arise during that construction. `this.contentWarnings`
-   * will contain all such errors, in the order they have been encountered.
-   * @param {Object} parsedManifest
-   * @param {Object} options
-   */
-  constructor(parsedManifest : IParsedManifest, options : IManifestParsingOptions) {
-    super();
-    const { supplementaryTextTracks = [],
-            supplementaryImageTracks = [],
-            representationFilter,
-            manifestUpdateUrl } = options;
-    this.contentWarnings = [];
-    this.id = generateNewManifestId();
-    this.expired = parsedManifest.expired ?? null;
-    this.transport = parsedManifest.transportType;
-    this.clockOffset = parsedManifest.clockOffset;
-
-    this.periods = parsedManifest.periods.map((parsedPeriod) => {
-      const period = new Period(parsedPeriod, representationFilter);
-      this.contentWarnings.push(...period.contentWarnings);
-      return period;
-    }).sort((a, b) => a.start - b.start);
-
-    /**
-     * @deprecated It is here to ensure compatibility with the way the
-     * v3.x.x manages adaptations at the Manifest level
-     */
-    /* eslint-disable import/no-deprecated */
-    this.adaptations = this.periods[0] === undefined ? {} :
-                                                       this.periods[0].adaptations;
-    /* eslint-enable import/no-deprecated */
+  return manifestObject;
 
-    this._timeBounds = parsedManifest.timeBounds;
-    this.isDynamic = parsedManifest.isDynamic;
-    this.isLive = parsedManifest.isLive;
-    this.isLastPeriodKnown = parsedManifest.isLastPeriodKnown;
-    this.uris = parsedManifest.uris === undefined ? [] :
-                                                    parsedManifest.uris;
-
-    this.updateUrl = manifestUpdateUrl;
-    this.lifetime = parsedManifest.lifetime;
-    this.suggestedPresentationDelay = parsedManifest.suggestedPresentationDelay;
-    this.availabilityStartTime = parsedManifest.availabilityStartTime;
-    this.publishTime = parsedManifest.publishTime;
-    if (supplementaryImageTracks.length > 0) {
-      this._addSupplementaryImageAdaptations(supplementaryImageTracks);
-    }
-    if (supplementaryTextTracks.length > 0) {
-      this._addSupplementaryTextAdaptations(supplementaryTextTracks);
-    }
-  }
-
-  /**
-   * Returns the Period corresponding to the given `id`.
-   * Returns `undefined` if there is none.
-   * @param {string} id
-   * @returns {Object|undefined}
-   */
-  public getPeriod(id : string) : Period | undefined {
-    return arrayFind(this.periods, (period) => {
-      return id === period.id;
+  /** @link IManifest */
+  function getPeriod(periodId : string) : IPeriod | undefined {
+    return arrayFind(manifestObject.periods, (period) => {
```

Comment:
Hi,

> not only does this allocate more function objects

Speaking strictly in  terms of `IManifest`, there's usually a single object at a time - at least in the majority of cases where there's only a single video player on screen.

Though you could make the same point for subparts of an `IManifest` like `IPeriod`, `IAdaptation` and so on.  Though even here, I 'm unsure that this is a real issue, at least in terms of memory imprint and performance - if that's what you mean.
To test.

> all call sites that invoke IManifest methods become polymorphic when the second IManifest object is created

Here, I'm not sure of what you exactly mean (in terms of "what problem does it create").
How I get it is: calling those methods will go, in the JS engine, through some sort of dynamic dispatch to get to the right function whereas having a single external function could unleash performance and memory optimizations inside the JS engine by allowing static dispatch to an uniquely-declared function. Is that it?

If that's so, I'm unsure that this is a real problem in the end, we're not at that level of optimization for now and we favored readability and simplicity (more information on why we do not rely on the alternatives below).

>  Also note that this approach will make the methods work when called without the proper this

This is taken into account, as there's no reason to use `this` in that code. In those inner functions, we're neither inside a class nor inside a well-defined JavaScript Object. That's the reason we're relying on `manifestObject` instead.
So I'm unsure if that's an issue here?

---

For the alternatives:

> Continue to define a Manifest class but only export a function that returns IManifest.

The point was to remove the idea of the Manifest being defined as a class to simplify several points: allowing asynchronous construction, complex return values while constructing without impacting the original object, and allowing several constructors. I will also add now: simplify cross-worker communication of the Manifest object.

So here, the fact that the `IManifest` type exists instead of the `Manifest` class is more of a (sad) side-effect of what we want to do.

> Define the methods at the top level of the file like function getPeriod(this : IManifest, periodId : string) : IPeriod | undefined and only use this to refer to the IManifest object.

We though about some versions of this, but this creates an usability issue when actually calling it:

  - either you import and call the function directly at call sites, where it's much less convenient than just relying on methods (it may though be the solution if we ever end up relying on webworker, like we plan to long term)

  - either you do things like bind the `this` inside the `createManifestObject` function, but there I'm unsure about the readability of the thing compared to what we have here.
    Again, I'm not sure that the performance issue you're noticing are big enough to being taken care of? I may be wrong though. That solution could be taken if we do see an issue with how this is actually developed, of in the future again if we want to do some sort of cross-worker whole manifest communication.

Reply:
> > not only does this allocate more function objects

> Speaking strictly in terms of `IManifest`, there's usually a single object at a time - at least in the majority of cases where there's only a single video player on screen.

> Though you could make the same point for subparts of an `IManifest` like `IPeriod`, `IAdaptation` and so on. Though even here, I 'm unsure that this is a real issue, at least in terms of memory imprint and performance - if that's what you mean. To test.

> > all call sites that invoke IManifest methods become polymorphic when the second IManifest object is created

> Here, I'm not sure of what you exactly mean (in terms of "what problem does it create"). How I get it is: calling those methods will go, in the JS engine, through some sort of dynamic dispatch to get to the right function whereas having a single external function could unleash performance and memory optimizations inside the JS engine by allowing static dispatch to an uniquely-declared function. Is that it?

Exactly.

> If that's so, I'm unsure that this is a real problem in the end, we're not at that level of optimization for now and we favored readability and simplicity (more information on why we do not rely on the alternatives below).

My point is that "JS where methods have one function object per object" is a slower language than "JS where methods are shared between objects of a given type". Replacing a single class with this new mechanism may not have a measurable effect, but doing it in a major way is likely to slow down execution uniformly (that is, the slowdown will not appear as a hot spot).

Callbacks will often be polymorphic or megamorphic, and that's OK. The point is to avoid it where easily possible, in the general way of writing code.

See for example https://www.builder.io/blog/monomorphic-javascript .

Many of the blog posts on Google have the fallacy that fixing hot spots is all you need for good performance, though.

> > Also note that this approach will make the methods work when called without the proper this

> This is taken into account, as there's no reason to use `this` in that code. In those inner functions, we're neither inside a class nor inside a well-defined JavaScript Object. That's the reason we're relying on `manifestObject` instead. So I'm unsure if that's an issue here?

My point was that, if such objects are exposed to the outer world and the methods work without `this`, external code will start to depend on it and it will be hard to go back.

> For the alternatives:

> > Continue to define a Manifest class but only export a function that returns IManifest.

> The point was to remove the idea of the Manifest being defined as a class to simplify several points: allowing asynchronous construction, complex return values while constructing without impacting the original object, and allowing several constructors. I will also add now: simplify cross-worker communication of the Manifest object.

To deal with this kind of thing (except cross-worker communication), I've often marked the constructor private and created static factory methods for outside code to create objects. The factory method can be async, can return additional information and there can be multiple of them.

> So here, the fact that the `IManifest` type exists instead of the `Manifest` class is more of a (sad) side-effect of what we want to do.

> > Define the methods at the top level of the file like function getPeriod(this : IManifest, periodId : string) : IPeriod | undefined and only use this to refer to the IManifest object.

> We though about some versions of this, but this creates an usability issue when actually calling it:

>     * either you import and call the function directly at call sites, where it's much less convenient than just relying on methods (it may though be the solution if we ever end up relying on webworker, like we plan to long term)

Functions acting on pure data are an option there, but indeed it is harder to find those functions. Unfortunately, a feature like extension methods would be against the goals of TypeScript.

An alternative could be to wrap a class around the pure data that can be transferred across workers.

>     * either you do things like bind the `this` inside the `createManifestObject` function, but there I'm unsure about the readability of the thing compared to what we have here.

What I was thinking of was to assign the function (which internally uses `this` to access the `IManifest`) directly, so just like with a class it requires a proper `this` to be passed and is the same for all `IManifest` objects. The function would not be exported directly.

`Function.prototype.bind` tends to be slow both when creating the bound function and when calling it, especially on older devices, so I would not recommend that.

>       Again, I'm not sure that the performance issue you're noticing are big enough to being taken care of? I may be wrong though. That solution could be taken if we do see an issue with how this is currently developed, of in the future again if we want to do some sort of cross-worker whole manifest communication.

I must admit I have not measured this ;-)

Reply:
OK I understand why that may be an issue (instinctively I would fear more the memory build-up of having a lot of completely different closures for each `Representation` of a content than the potential perf issues when calling those - at least in the way the RxPlayer is currently written).

In this PR - that probably won't get merged as it is outdated anyway (though is kept alive as a reference) - I was trying to see what a Manifest object created by a function instead of by a class would look like, for the aforementioned reasons. Through the few tests performed, I did not see any sensible performance or memory issues - though we're lately experimenting with much larger Manifests due to ad-switched live contents for example, even sometimes with several codecs, language and accessibility choices, so it's possible that it may become a real issue in the future.

> To deal with this kind of thing (except cross-worker communication), I've often marked the constructor private and created static factory methods for outside code to create objects. The factory method can be async, can return additional information and there can be multiple of them.

Here we cannot do this because it's the the instantiation of a sub-class (the `Representation` which is itself in the `Adaptation` class, itself in the `Period` class, itself in the `Manifest` class) whose construction might become asynchronous - so even the `Manifest`'s has to become asynchronous under that architecture - which is not possible.

Though we have other tricks up our sleeves - such as starting the asynchronous task at subscription and reflect its advancement through one of those `Representation`'s property or just perform the asynchronous task elsewhere later. All of them will keep construction synchronous.

> What I was thinking of was to assign the function (which internally uses this to access the IManifest) directly, so just like with a class it requires a proper this to be passed and is the same for all IManifest objects. The function would not be exported directly.

I see. Why not if we ever actually continue down this path.

I did not play enough with TypeScript to know how to communicate what `this` has to be - nor if it can enforce it - I'll check.

---

Note that for now, I just keep the class format until it appears very un-adapted. Then we'll see.

For example, I recently worked-around the issue of a constructor not being able to return other values (the idea was to recuperate warnings seen at construction) by just declaring a `warnings` parameter on the constructor (which as to be set as an array - then filled by warnings seen at constructions) [in a recent PR](https://github.com/canalplus/rx-player/pull/1264).
Though this isn't idiomatic much, but that's only a minor inconvenience for now.

For the asynchronous part, I was mainly thinking about using the [MediaCapabilities API](https://developer.mozilla.org/en-US/docs/Web/API/Navigator/mediaCapabilities) at construction to report a better support analysis of each `Representation` (a.k.a. media qualities) seen in the Manifest. As the main method here returns a Promise.
Though I'm still PoCing that thing and there's multiple possible solutions that could still let us rely on a class.

Lastly for the worker part, the issue is that we need some form of the Manifest in both the main and worker JS realms. But worker <-> main communication is mostly done by [`postMessage`](https://developer.mozilla.org/en-US/docs/Web/API/Worker/postMessage)  calls until now. For various reasons, we cannot communicates the full class (more specifically its methods) through that mean, only serializable properties.

Though there's also other differences we want to keep here (for example, the main thread doesn't need to know the list of available segments whereas the worker does) and we might also want to limit properties mutations (as the effect of the mutation wouldn't be automatically "seen" on the other side), so we still need to experiment a lot here.